### PR TITLE
Fixing bugs with some GET and TTL update paths

### PR DESCRIPTION
**Pull request by vgkholla**
1. Calls TTL update only if the GET succeeded
2. Records server level errors on `GetBlobOperation` and `GetBlobInfoOperation`

### CODE REVIEW

File: ambry-router/src/main/java/com.github.ambry.router/GetBlobInfoOperation.java [hunk 1]
```diff
@@ -311,6 +311,7 @@ private void processGetBlobInfoResponse(GetRequestInfo getRequestInfo, GetRespon
     } else {
       logger.trace("Replica {} returned an error {} for a GetBlobInfoRequest with response correlationId : {} ",
           getRequestInfo.replicaId.getDataNodeId(), getError, getResponse.getCorrelationId());
+      processServerError(getError);
```
Comment (cgtz) [inline ambry-router/src/main/java/com.github.ambry.router/GetBlobInfoOperation.java]: So the old code would only consider partition level error and calling `processServerError` here will ensure that server-level errors (like Disk_Unavailable) will be tracked by `ResponseHandler`?

File: ambry-router/src/test/java/com.github.ambry.router/GetBlobInfoOperationTest.java [hunk 2]
```diff
@@ -414,6 +390,26 @@ public void testBlobAuthorizationFailureOverrideAll() throws Exception {
...
 
+  /**
+   * Tests the case where all servers return the same server level error code
+   * @throws Exception
+   */
+  @Test
+  public void testFailureOnServerErrors() throws Exception {
+    List<ServerErrorCode> serverErrors = new ArrayList<>(Arrays.asList(ServerErrorCode.values()));
+    // set the status to various server level level errors (remove all partition level errors or non errors)
+    serverErrors.remove(ServerErrorCode.Blob_Deleted);
```
Comment (cgtz) [inline ambry-router/src/test/java/com.github.ambry.router/GetBlobInfoOperationTest.java]: Could also use `EnumSet.complementOf(EnumSet.of(Blob_Expired, No_Error, ...))`

File: ambry-router/src/test/java/com.github.ambry.router/GetBlobOperationTest.java [hunk 3]
```diff
@@ -467,17 +467,40 @@ public void testSuccessInThePresenceOfVariousErrors() throws Exception {
...
+  }
+
+  /**
+   * Tests the case where all servers return the same server level error code
+   * @throws Exception
+   */
+  @Test
+  public void testFailureOnServerErrors() throws Exception {
+    doPut();
+    List<ServerErrorCode> serverErrors = new ArrayList<>(Arrays.asList(ServerErrorCode.values()));
```
Comment (cgtz) [inline ambry-router/src/test/java/com.github.ambry.router/GetBlobOperationTest.java]: Can use `EnumSet` here too, unless the order is important.

File: ambry-router/src/test/java/com.github.ambry.router/GetBlobInfoOperationTest.java [hunk 4]
```diff
@@ -414,6 +390,26 @@ public void testBlobAuthorizationFailureOverrideAll() throws Exception {
...
   }
 
+  /**
+   * Tests the case where all servers return the same server level error code
+   * @throws Exception
+   */
+  @Test
+  public void testFailureOnServerErrors() throws Exception {
+    List<ServerErrorCode> serverErrors = new ArrayList<>(Arrays.asList(ServerErrorCode.values()));
+    // set the status to various server level level errors (remove all partition level errors or non errors)
```
Comment (jsjtzyy) [inline ambry-router/src/test/java/com.github.ambry.router/GetBlobInfoOperationTest.java]: minor: please remove extra `level` in comment.

File: ambry-router/src/test/java/com.github.ambry.router/GetBlobInfoOperationTest.java [hunk 5]
```diff
@@ -414,6 +390,26 @@ public void testBlobAuthorizationFailureOverrideAll() throws Exception {
...
+    serverErrors.remove(ServerErrorCode.Blob_Deleted);
+    serverErrors.remove(ServerErrorCode.Blob_Expired);
+    serverErrors.remove(ServerErrorCode.No_Error);
+    serverErrors.remove(ServerErrorCode.Blob_Authorization_Failure);
+    serverErrors.remove(ServerErrorCode.Blob_Not_Found);
+    for (ServerErrorCode serverErrorCode : serverErrors) {
+      mockServerLayout.getMockServers().forEach(server -> server.setServerErrorForAllRequests(serverErrorCode));
+      assertOperationFailure(serverErrorCode == ServerErrorCode.Disk_Unavailable ? RouterErrorCode.AmbryUnavailable
```
Comment (jsjtzyy) [inline ambry-router/src/test/java/com.github.ambry.router/GetBlobInfoOperationTest.java]: question: if `Replica_Unavailable` is introduced, this type of server error should be mapped to `RouterErrorCode.AmbryUnavailable` as well?

File: ambry-router/src/test/java/com.github.ambry.router/GetBlobInfoOperationTest.java [hunk 6]
```diff
@@ -523,82 +496,66 @@ private void testVariousErrors(String dcWherePutHappened) throws Exception {
...
-      op.handleResponse(responseInfo, getResponse);
-      if (op.isOperationComplete()) {
-        break;
-      }
-    }
-
-    if (!op.isOperationComplete()) {
-      Assert.assertTrue("Latch should have been zeroed ", onPollLatch.await(500, TimeUnit.MILLISECONDS));
+  /**
+   * Coompletes {@code op}
```
Comment (jsjtzyy) [inline ambry-router/src/test/java/com.github.ambry.router/GetBlobInfoOperationTest.java]: minor typo here

File: ambry-router/src/main/java/com.github.ambry.router/GetBlobInfoOperation.java (hunk 1 above)
Comment (vgkholla) [reply ambry-router/src/main/java/com.github.ambry.router/GetBlobInfoOperation.java]: Correct. For instance, if all the servers returned `Disk_Unavailable` or `Unknown_Error`, the current code will not return any response

File: ambry-router/src/test/java/com.github.ambry.router/GetBlobInfoOperationTest.java (hunk 2 above)
Comment (vgkholla) [reply ambry-router/src/test/java/com.github.ambry.router/GetBlobInfoOperationTest.java]: Done

File: ambry-router/src/test/java/com.github.ambry.router/GetBlobOperationTest.java (hunk 3 above)
Comment (vgkholla) [reply ambry-router/src/test/java/com.github.ambry.router/GetBlobOperationTest.java]: It is not - done

File: ambry-router/src/main/java/com.github.ambry.router/GetBlobInfoOperation.java [hunk 7]
```diff
@@ -400,6 +401,9 @@ private void processServerError(ServerErrorCode errorCode) {
         logger.trace("Requested blob was not found on this server");
         setOperationException(new RouterException("Server returned: " + errorCode, RouterErrorCode.BlobDoesNotExist));
         break;
+      case Disk_Unavailable:
+        logger.trace("Requested blob is not accessible");
```
[the remark this answers is not in the document]
Comment (vgkholla) [reply ambry-router/src/main/java/com.github.ambry.router/GetBlobInfoOperation.java]: Done

File: ambry-router/src/test/java/com.github.ambry.router/GetBlobInfoOperationTest.java (hunk 4 above)
Comment (vgkholla) [reply ambry-router/src/test/java/com.github.ambry.router/GetBlobInfoOperationTest.java]: Done

File: ambry-router/src/test/java/com.github.ambry.router/GetBlobInfoOperationTest.java (hunk 6 above)
Comment (vgkholla) [reply ambry-router/src/test/java/com.github.ambry.router/GetBlobInfoOperationTest.java]: Fixed

File: ambry-router/src/test/java/com.github.ambry.router/GetBlobInfoOperationTest.java (hunk 5 above)
Comment (vgkholla) [reply ambry-router/src/test/java/com.github.ambry.router/GetBlobInfoOperationTest.java]: Yeah - I think both should map to `AmbryUnavailable`

**Comment by vgkholla** [comment]
@cgtz @jsjtzyy Comments addressed.

**Pull request merged**